torch/_inductor/utils.py

```diff
@@ -3591,31 +3591,46 @@ def python_subprocess_env() -> dict[str, str]:
     return env
 
 
-# A tuple of customized partition wrappers from users. Inductor mechanically wraps
-# each partition fn with the customized partition wrapper. When using for CUDAGraph wrapper,
-# users need to handle all details such as static inputs, dynamic shapes, etc.
-# Assume there is only 1 dynamo graph in the process and the number of partitions
-# equals to the number of customized partition wrappers.
-#
-# Warning: This API is unstable and may change in the future.
-_unstable_customized_partition_wrappers: list[Callable[..., Callable[..., Any]]] = []
-
-
-def set_customized_partition_wrappers(
-    wrappers: tuple[Callable[..., Callable[..., Any]], ...],
-    metadata: Optional[tuple[Any, ...]] = None,
-) -> None:
+@dataclasses.dataclass(frozen=True)
+class CUDAGraphWrapperMetadata:
     """
-    Set customized partition wrappers for the current process.
+    Metadata for Customized CUDAGraphWrapper.
 
-    Args:
-        wrappers (tuple[Callable[..., Callable[..., Any]]]): A tuple of
-            customized partition wrappers.
-        metadata (tuple[Any]): Metadata for the partition wrappers.
+    Currently assumes there is 1 dynamo graph and will extend to
+    multiple graphs in the future.
     """
-    # metadata is unused for now
-    assert metadata is None
-    global _unstable_customized_partition_wrappers
 
-    _unstable_customized_partition_wrappers.clear()
-    _unstable_customized_partition_wrappers.extend(wrappers)
+    # The number of partitions that are cudagraphable.
+    num_partitions: int
+
+    # The number of partitions that are cudagraphable and have a fallback.
+    partition_index: int
+
+
+PartitionFnType = Callable[..., Any]
+CUDAGraphWrapperType = Callable[
+    [PartitionFnType, CUDAGraphWrapperMetadata], PartitionFnType
+]
+
+
+# only incremented by user call of mark_step_begin
+class CUDAGraphWrapper:
+    wrapper: Optional[CUDAGraphWrapperType] = None
```

Comment:
Not sure why this is nested but I'm not opposed

Reply:
only for variable scope purpose